### PR TITLE
Filter widgets by Culture - Fixes #4961

### LINKED ISSUE #4961: Localizing widgets on layers
I cannot find a way to localize widgets that are contained in layers, I've performed the following: -
1. Add LocalizablePart to Liquid widget
2. Re-save the widget and add the localizable contents

At this point I just see multiple copies in the front end.

To get around this I have modified LayerService to de-dup based on the current culture

```csharp
        public async Task<IEnumerable<ContentItem>> GetLayerWidgetsAsync(
            Expression<Func<ContentItemIndex, bool>> predicate)
        {
            var widgets = await _session
                .Query<ContentItem, LayerMetadataIndex>()
                .With(predicate)
                .ListAsync();

            widgets = (await _contentLocalizationManager.DeduplicateContentItemsAsync(widgets)).Values;

            return widgets;
        }
```

and in LayerFilter, changed the cache key to include the culture
```csharp
                var widgets = await _memoryCache.GetOrCreateAsync($"OrchardCore.Layers.LayerFilter:AllWidgets-{Thread.CurrentThread.CurrentUICulture}", entry =>
                {
                    entry.AddExpirationToken(_signal.GetToken(LayerMetadataHandler.LayerChangeToken));
                    return _layerService.GetLayerWidgetsMetadataAsync(x => x.Published);
                });
```

Is there a better way to handle widget localization?

### PR DESCRIPTION
**Pull request by PaulEpitomy**
I've implemented this using the culture aspect as it was much simpler that way. 

The one area that I'm unsure about is the GraphQL API which returns widgets, this PR will filter those by culture, how should this behave?

Fixes #4961

### CODE REVIEW

**Comment by Skrypt** [comment]
Yeah looking at your PR and it seems that you changed the main method that fetch the widgets to make it filter with Localization but you need to take in consideration those that don't require this. So, you fixed one issue but you should instead also keep what was there. It needs to work for both scenarios where a LocalizationPart would be attached to a Widget and when there's not.

**Comment by PaulEpitomy** [comment]
I've moved the culture filtering into the layer filter, so as to not impact the other uses of the service. Using the CultureAspect already caters for the case where no LocalizationPart is attached to the ContentItem.

File: src/OrchardCore.Modules/OrchardCore.Layers/Services/LayerFilter.cs [hunk 1]
```diff
@@ -146,5 +152,18 @@ public async Task OnResultExecutionAsync(ResultExecutingContext context, ResultE
 
             await next.Invoke();
         }
+
+        private async Task<IEnumerable<LayerMetadata>> FilterWidgetsByCultureAsync(IEnumerable<LayerMetadata> widgets)
+        {
+            var widgetsWithCulture = await Task.WhenAll(widgets.Select(async w =>
+            {
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Layers/Services/LayerFilter.cs]: Please don't use `Task.WhenAll()` when it triggers the execution of scoped services.

File: src/OrchardCore.Modules/OrchardCore.Layers/Services/LayerFilter.cs [hunk 2]
```diff
@@ -146,5 +152,19 @@ public async Task OnResultExecutionAsync(ResultExecutingContext context, ResultE
...
             await next.Invoke();
         }
+
+        private async Task<IEnumerable<LayerMetadata>> FilterWidgetsByCultureAsync(IEnumerable<LayerMetadata> widgets)
+        {
+            var orchardHelper = _orchardHelper;
+            var widgetsWithCulture = await Task.WhenAll(widgets.Select(async w =>
+            {
+                var culture = await orchardHelper.GetContentCultureAsync(w.ContentItem);
+                return new { Widget = w, Culture = culture };
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Layers/Services/LayerFilter.cs]: I don't think this anonymous object is necessary. You can use IAsyncEnumerable instead by iterating over each widget.

**Comment by sebastienros** [comment]
The next thing we'll need is a dropdown in the layers page to filter which culture to display, in case we want to see what widgets are for a specific culture. Should be easy to do, with the same technique that is used in this PR to not render in the front-end.

**Comment by hishamco** [comment]
@PaulEpitomy any updates on the PR?

**Comment by PaulEpitomy** [comment]
@hishamco Sorry thought I'd already pushed these changes.

**Comment by Skrypt** [comment]
871 files changed? Please merge dev branch on top of yours and fix merge issues. Else ask for help from someone.

**Comment by Skrypt** [comment]
Ok makes more sense now

File: src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersOptions.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.Rendering;
+
+namespace OrchardCore.Layers.ViewModels
+{
+    public class LayersOptions
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersOptions.cs]: LayersOptionsViewModel

File: src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersOptionsViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.Rendering;
+
+namespace OrchardCore.Layers.ViewModels
+{
+    public class LayersOptionsViewModel
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersOptionsViewModel.cs]: I'm sorry I removed my comment and you did not see it. I'm not sure this is the appropriate way to do this.

File: src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersIndexViewModel.cs [hunk 5]
```diff
@@ -14,5 +14,7 @@ public class LayersIndexViewModel
      
         [BindNever]
         public List<Layer> Layers { get; set; }
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersIndexViewModel.cs]: Here we have Layers which should have Options which you would define on the Layer.cs class itself?

File: src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersIndexViewModel.cs [hunk 6]
```diff
@@ -14,5 +14,7 @@ public class LayersIndexViewModel
...
+
+        public LayersOptionsViewModel Options { get; set; } = new LayersOptionsViewModel();
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersIndexViewModel.cs]: Bacause this is not a ViewModel but a Model.

File: src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersIndexViewModel.cs (hunk 5 above)
Comment (PaulEpitomy) [reply src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersIndexViewModel.cs]: Not sure it does? Are you suggesting adding options to the Layer class? At present the Options are providing a list of available cultures and a selected culture, what need does the Layer class have of that or have I misunderstood?

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Layers/ViewModels/LayersIndexViewModel.cs]: Keep it like that I will analyze further later.

**Comment by agriffard** [comment]
Can you please fix the conflicts and do the suggested changes?

**Comment by MikeKry** [comment]
Can I help somehow with this PR? 

This change is must have for multilang sites (if we want to use layers).

**Comment by Piedone** [comment]
Is this something you'd like to revisit @PaulEpitomy any time soon or should we close? @MikeKry if you can work on this then please feel free to take the current code, address the above-mentioned issues, and open a new PR, that would be awesome.

**Comment by hishamco** [comment]
@MikeKry are you still interested on this?

**Comment by MikeKry** [comment]
@hishamco No, we have stopped using layout widgets. It was not very good for headless recipe anyway.

**Comment by Piedone** [comment]
Closing it, then.

**Comment by panayiotisp** [comment]
Was this ever handled? i.e. to have localized widgets in the layers

**Comment by Piedone** [comment]
Not in exactly this way, but you can have layers that match specific cultures in their layer rules.

**Comment by hishamco** [comment]
I might open a PR that shows a layer for specific culture(s) as @Piedone suggested which is make sense to me

**Comment by Piedone** [comment]
That already exists, the Culture condition is for that.

**Comment by hishamco** [comment]
Maybe I'm not aware about it, thanks anyway

**Comment by panayiotisp** [comment]
Thanks both. I was already using the culture in the rules (in combination with the URL), just wanted to see if there was something more inline with the other localization patterns in orchard core as we will add a few cultures/languages to the module we are working on.

But indeed the layer works fine without any issues

**Comment by hishamco** [comment]
@Piedone is there any docs for layers and cultures?

**Comment by Piedone** [comment]
Did you search for it, Hisham? :)

**Pull request closed without merging**